### ISSUE TITLE
🦠 Model Request: Cyp2c9 Substrate

### ISSUE THREAD

**Issue by GemmaTuron** [ersilia-os/ersilia#431]
### Model Title

CYP2C9 Substrate (TDC dataset)

### Publication

Hello @thormiwa!

As part of your Outreachy contribution, we have assigned you the dataset "cyp2c9 substrate" from the Therapeutics Data Commons to try and build a binary classification ML model.
Please copy the provided Google Colab template and use this issue to provide updates on the progress. We'll value not only being able to build the model but also interpreting its results.

**Comment by thormiwa** [comment]
@GemmaTuron Thanks if i have any questions i will reach out :)

**Comment by GemmaTuron** [comment]
YEs @thormiwa !

Please make this clarification on the slack channel as well so that no one gets confused ;)

**Comment by thormiwa** [comment]
Hi @GemmaTuron I understand that contingency matrix compares the true value to the predicted value, but looking at the true value it is a binary list with 1, 0....but the predict value is more like a numpy array with 0.89580, 0.11849 for example....how exactly can the confusion matrix compare them correctly.

**Comment by ZakiaYahya** [comment]
@thormiwa you can do it by using round( ) function to map values to 1 and 0.

**Comment by GemmaTuron** [comment]
Hi @thormiwa 

The confusion matrix functions sometimes automatically convert the probabilities to binary (0 or 1) using a cut-off of 0.5 (depends on the function you have selected). I find it best to use a for loop to convert my array to a list of 0 and 1 with the threshold I want (not always 0.5)

**Comment by thormiwa** [comment]
Hi @GemmaTuron You mentioned in the meeting that we could save our models in a joblib format so we don't have to run over and over again..is there a way to use the model that we saved since you have to re run codes on colab everytime we go away from the colab code.

**Comment by Malikbadmus** [comment]
@thormiwa , yes you can.

Use the joblib.load function to load your model from your drive.

**Comment by thormiwa** [comment]
Hi @GemmaTuron I will be explaining about the steps i took to load the dataset and every other step in understanding the model assigned to me.

**Comment by thormiwa** [comment]
### **Load your Data**

The model assigned is the **CYP2C9 Substrate, Carbon-Mangels et al.** 
CYP P450 2C9 plays a major role in the oxidation of both xenobiotic and endogenous compounds including fatty acids. In humans, the protein is encoded by the CYP2C9 gene. Substrates are drugs that are metabolized by the enzyme.

From TDCOMMONS, The task description is to predict if a drug SMILES string is a substrate to an enzyme.
The Dataset has 666 Drugs

### Split Your Data
Using the tdcommons get_split function, the dataset can be splitted into the Train Data, Validation Data and the Test Data

- **Train Data** The Train Dataset is Set of data used for learning (by the model), that is, to fit the parameters to the machine learning model
- **Validation Data** The Validation Dataset is an unbiased set of data that is used to improve the model. it can also play a major role when selecting a threshold for the model.
- **Test Data** The Test Dataset can be used to provide unbiased evaluation the final performance of a model.


### Analyze Your Data
In the CYP2C9 Substrate dataset, there are

- There are 468 molecules in the train dataset. 97 active and 371 inactive
- There are 67 molecules in the valid dataset. 17 active and 50 inactive
- There are 134 molecules in the test dataset. 27 active and 107 inactive

From the explanation above, we can see we have an **Imbalanced Dataset**

There are several ways to Visualize your dataset. I started by using the Matplotlib Python Package.
i preferred to use a BarPlot to plot the data as it is more readable since we are only seeing two outcomes i.e Active or Inactive and it takes numerical values.

![matplotlib](https://user-images.githubusercontent.com/70862950/198087965-041fbc86-fc17-4f33-9aec-b5ce92a46052.png)

I also used the RDKIT package to visualise and have an image of what an active molecule looks like and an inactive molecule. The rdkit package is useful for data visualisation as it can draw chemical Molecules.

**This is an example of an Active molecule**

![active_mols](https://user-images.githubusercontent.com/70862950/198091431-6a0d26d7-1cc9-4780-8ba4-3e591ceed709.png)

**This is an example of an Inactive molecule**
![inactive_mols](https://user-images.githubusercontent.com/70862950/198091517-bcf21af9-8446-45fb-8eb9-575801a0a05f.png)

**Comment by thormiwa** [comment]
### **Training The Model**

Training a model simply means learning (determining) good values for all the weights and the bias from labeled examples. The training data must contain the correct values i.e the target attributes. 
For example, With the **Cyp2c9 Substrate**  we are to predict if a drug is a subtrate to the enzyme. 
The training data that contains drugs for which you know the target (that is, a label that tells whether a drug is a subtrate to an enzyme or not). we would train the ML model by using this data, resulting in a model that attempts to predict whether it is a subtrate or not.

I started by installing the **Lazy QSAR** Package. With this package we can train (and validate) a model based on Morgan fingerprints. With the help of the Lazy QSAR package our smiles string can be converted to a vector unless the computer won't be able to understand them. Each Vector( formerly smiles) will be associated with an outcome. Our ML will try to get familiar with the vector since it can read them , on what is an active or inactive so if we come with a new molecule it can tell us if it is a subtrate to the enzyme(active) or not. The time budget given to the fitting process actually matters as it help the ML to train the data. It is also smart to be limited with this time specified. It is also good to save the model. I saved the model in the joblib format, Which can be reuseable.

The next thing is to Predict using the model trained.  Using the validation set to evaluate results from the training set. 
The validation set is a new set and it will be an unbiased data as the model hasn't trained it. This process involves having a list of the smiles drug in the validation set and the outcome of the validation set, we pass these list to our model to get predictions and compare it to our Actual true value.

**Comment by thormiwa** [comment]
### **Evaluate your model**

it's important to use new data when evaluating our model to prevent the likelihood of overfitting to the training set.

### Classification metrics
When performing classification predictions, there's four types of outcomes that could occur.

- **True positives** are when you predict an observation belongs to a class and it actually does belong to that class. In our case when a smiles drug is a subtrate to an enzyme and the model predicts it as a subtrate to an enzyme

- **True negatives** are when you predict an observation does not belong to a class and it actually does not belong to that class.
In our case when a smiles drug is not a subtrate to an enzyme and the model predicts it as not subtrate to an enzyme.

- **False positives** occur when you predict an observation belongs to a class when in reality it does not. In our case when a smiles drug is not a subtrate to an enzyme and the model predicts it as a subtrate to an enzyme

- **False negatives** occur when you predict an observation does not belong to a class when in fact it does. In our case when a smiles drug is a subtrate to an enzyme and the model predicts it as not subtrate to an enzyme.

**Comment by thormiwa** [comment]
@GemmaTuron  This is a link to my Colab Notebook.
https://colab.research.google.com/drive/1J682m9GhulXqKzuScAxi5BU7RncwEhxy?usp=sharing

**Comment by GemmaTuron** [comment]
Hi @thormiwa !

Very good job and interpretation work. I would suggest increasing model training time to improve the AUC that you are getting, which is not great (less than 0.65)

**Comment by thormiwa** [comment]
> Hi @thormiwa !
> 
> Very good job and interpretation work. I would suggest increasing model training time to improve the AUC that you are getting, which is not great (less than 0.65)

HI @GemmaTuron Thank you for your suggestion i would increase the time and update the metrics used.

**Comment by Zainab-ik** [comment]
> > Hi @thormiwa !
> > Very good job and interpretation work. I would suggest increasing model training time to improve the AUC that you are getting, which is not great (less than 0.65)
> 
> HI @GemmaTuron Thank you for your suggestion i would increase the time and update the metrics used.

@thormiwa, you can try training your model for 10 minutes up to 30 minutes and also try different learning algorithms to see which performed better.

**Comment by thormiwa** [comment]
Hi @GemmaTuron I have updated the AUC after model training time like you suggested. If everything is okay you can give me the go ahead to close the issue. :)

**Comment by GemmaTuron** [comment]
Hi @thormiwa !

Thanks for the good job, I'll close the issue and you can go onto finalising your outreachy application
